File: ios/CleanArchitecture/Login/LoginPresenter.swift
//
//  LoginPresenter.swift
//  CleanArchitecture
//
//  Created by Alan Donizete Quintiliano on 12/04/19.
//  Copyright © 2019 Alan Quintiliano. All rights reserved.
//

import UIKit
import main

class LoginPresenter: BasePresenter, CallBack {
    
    var view: LoginView?
    let useCase: LoginUseCase
    
    init(useCase: LoginUseCase) {
        self.useCase = useCase
    }
    
    func onCancel() {
        
    }
    
    func attach(view: BaseView) {
        self.view = view as? LoginView
    }
    
    func onClickLoginButton() {
        view?.showProgress()
        LoginController(loginUseCase: self.useCase, callback: self)
            .attemptLogin(email: view?.getEmail(), password: view?.getPassword())
    }
    
    func onFinish(result: Any?) {
        view?.hideProgress()
    }
    
    func onError(error_ error: String) {
        view?.onFail(error: error)
    }
}

File: android/src/main/java/com/adqmobile/cleanarchitecture/login/LoginPresenter.kt
package com.adqmobile.cleanarchitecture.login

import com.adqmobile.controllers.BasePresenter
import com.adqmobile.controllers.BaseView
import com.adqmobile.controllers.LoginController
import com.adqmobile.controllers.LoginView
import com.adqmobile.controllers.task.CallBack
import com.adqmobile.domain.entities.LoginResponseEntity
import com.adqmobile.domain.usecases.LoginUseCase

class LoginPresenter(private val useCase: LoginUseCase): BasePresenter, CallBack<LoginResponseEntity> {

    var view: LoginView? = null

    override fun attach(view: BaseView) {
        this.view = view as? LoginView
    }

    fun onClickLoginButton() {
        view?.showProgress()
        LoginController(this.useCase, this)
            .attemptLogin(view?.getEmail(), view?.getPassword())
    }

    override fun onCancel() {

    }

    override fun onFinish(result: LoginResponseEntity?) {
        view?.hideProgress()
    }

    override fun onError(error: String) {
        view?.hideProgress()
        view?.onFail(error)
    }
}